src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc

```diff
@@ -0,0 +1,725 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <memory>
+#include <string_view>
+#include <iomanip> 
+
+#include "src/common/base/base.h"
+#include "src/stirling/utils/parse_state.h"
+#include "src/stirling/utils/binary_decoder.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/types.h"
+
+#define PX_ASSIGN_OR_RETURN_ERROR(expr, val_or) \
```

Comment:
Can we rename this to `PX_ASSIGN_OR_RETURN_NEEDS_MORE_DATA`? Imo `kNeedsMoreData` isn't an error because it's an indicator that we should continue to retry. We also have the same macro defined in the pgsql parser and this would ensure our naming is consistent ([source](https://github.com/pixie-io/pixie/blob/62b8080699371b0bcd83e2c3ddda265d22d0eb40/src/stirling/source_connectors/socket_tracer/protocols/pgsql/parse.cc#L41C28-L41C28)).

Reply:
So what that would mean is only in the main parsing function before parsing the variable header or the payload, checks are done to make sure remaining length is equal to the buffer size and then for any subsequent bit extractions, `kInvalid` should be returned? 
Right now for every extraction error `kNeedsMoreData` is being returned. I realise that extraction error happening is unlikely (unless remaining length field is incorrect) because we have already validated that the remaining length is equal to the buffer size so the full packet is already in the buffer. Would you suggest having PX_ASSIGN_OR_RETURN_INVALID to return `kInvalid` in variable header parsing and payload parsing? Because an extraction error in variable header/payload parsing, after we have already validated remaining length to be equal to buffer size, means that the remaining length field was wrong, which would make it an invalid MQTT packet.

Reply:
Your latest changes are consistent with what I described above, so this lgtm.